tempesta_fw/http_parser.c

```diff
@@ -1237,6 +1306,95 @@ __req_parse_x_forwarded_for(TfwHttpMsg *msg, unsigned char *data, size_t len)
 	return r;
 }
 
+static int
+__req_parse_user_agent(TfwHttpMsg *msg, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	TfwHttpParser *parser = &msg->parser;
+	unsigned char *p = data;
+	unsigned char c = *p;
+	__FSM_DECLARE_VARS();
+
+	__FSM_START(parser->_i_st) {
+
+	__FSM_STATE(Req_I_UserAgent) {
+		__fsm_sz = len - (size_t)(p - data);
+		__fsm_ch = memchr(p, '\r', __fsm_sz);
+		if (__fsm_ch)
+			return __fsm_ch - data;
+		__FSM_I_MOVE_n(Req_I_UserAgent, __fsm_sz);
+	}
+
+	} /* FSM END */
+
+done:
+	return r;
+}
+
+static int
+__req_parse_cookie(TfwHttpMsg *msg, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	TfwHttpParser *parser = &msg->parser;
+	unsigned char *p = data;
+	unsigned char *orig_data = data;
+	unsigned char c = *p;
+	__FSM_DECLARE_VARS();
+
+	/*
+	 * Here we build header value string manually
+	 * to split it in chunks: chunk bounds are
+	 * at least at name start, value start and value end.
+	 * This simplifies cookie search, http_sticky uses it.
```

Comment:
Please add a comment with particular RFC chapter which describes the parsing rules.